src/maxtree.jl

```diff
@@ -0,0 +1,859 @@
+
+    # create uninitialized MaxTree for image
+    MaxTree{N}(image::GenericGrayImage{<:Any, N}, rev::Bool) where N =
+        new{N}(rev, similar(image, Int),
+               Vector{Int}(undef, length(image)))
+end
+
+Base.ndims(::Type{<:MaxTree{N}}) where N = N
+Base.ndims(maxtree::MaxTree) = ndims(typeof(maxtree))
+Base.length(maxtree::MaxTree) = length(maxtree.parentindices)
+Base.size(maxtree::MaxTree) = size(maxtree.parentindices)
+Base.:(==)(a::MaxTree, b::MaxTree) =
+    (ndims(a) == ndims(b)) && (a.rev == b.rev) &&
+    (a.parentindices == b.parentindices) &&
+    (a.traverse == b.traverse)
+
+"""
+    root_index(maxtree::MaxTree) -> Int
+
+Linear index of the root pixel.
+"""
+root_index(maxtree::MaxTree) = maxtree.traverse[1]
+
+# Gets the path to the root of the max-tree (defined by `ancestors`) starting from
+# `node`. The indices of the nodes along the path are stored into `path`.
+# Returns the index of the root
+@inline function rootpath!(path::Vector{Int},
+                           ancestors::AbstractArray{<:Integer},
+                           node::Integer)
+    empty!(path)
+    p = node
+    @label next
+    push!(path, p)
+    @inbounds q = ancestors[p]
+    if p != q
+        p = q
+        @goto next
+    end
+    return p
+end
+
+# Corrects `maxtree.parentindices`, so that the parent of every pixel
+# is a canonical node. I.e. for each connected component at a level l,
+# all pixels point to the same representative, which in turn points to the
+# representative pixel at the next level.
+#
+# A canonical node ``p`` is either a root of the max-tree, or
+# ``image(parent(p)) != image(p)``
+function canonize!(maxtree::MaxTree, image::GenericGrayImage)
+    @assert size(maxtree) == size(image)
+    parentindices = maxtree.parentindices
+    @inbounds for p in maxtree.traverse
+        q = parentindices[p]
+        if image[q] == image[parentindices[q]]
+            parentindices[p] = parentindices[q]
+        end
+    end
+    return maxtree
+end
+
+# Checks whether the `pixel` + `offset` is still a valid pixel
+# contained inside the image area defined by `axes`.
+isvalid_offset(offset::CartesianIndex{N}, pixel::CartesianIndex{N},
+               axes::NTuple{N}) where N =
+    Base.checkbounds_indices(Bool, axes, (pixel + offset,))
+
+# generates a vector of offsets to the CartesianIndex that defines
+# the neighborhood of the pixel in an N-dimensional image.
+# `connectivity` is the maximal number of orthogonal steps
+# (+1/-1 to a single dimension) that are required to reach a neighbor.
+# Note that the corresponding offsets of linear indices would be sorted in ascending order
+# ensuring efficient memory access pattern.
+function neighbor_cartesian_offsets(::Type{CartesianIndex{N}}, connectivity::Integer) where N
+    (connectivity >= 1) || throw(ArgumentError("connectivity should be positive integer"))
+    ci1 = one(CartesianIndex{N})
+    return [ci for ci in -ci1:ci1 if 1 <= sum(abs, Tuple(ci)) <= connectivity]
```

Comment:
@timholy Julia 1.0 complains about `-ci1:ci1`:
```
MethodError: Cannot `convert` an object of type Int64 to an object of type CartesianIndex{1}
  Closest candidates are:
    convert(::Type{T}, !Matched::T) where T at essentials.jl:154
    CartesianIndex{1}(::Integer...) where N at multidimensional.jl:69
    CartesianIndex{1}(::Integer...) where N at multidimensional.jl:72
  Stacktrace:
   [1] oftype(::CartesianIndex{1}, ::Int64) at ./essentials.jl:334
   [2] (::Colon)(::CartesianIndex{1}, ::CartesianIndex{1}) at ./range.jl:7
   [3] neighbor_cartesian_offsets(::Type{CartesianIndex{1}}, ::Int64) at /home/travis/build/JuliaImages/ImageMorphology.jl/src/maxtree.jl:138
```
Is there a workaround?

Reply:
Similar to my previous comment:

```julia
# compat.jl
"""
    _colon(I:J)
`_colon(I, J)` works equivelently to `I:J`, it's used to backward support julia v"1.0".
"""
_colon(I, J) = I:J
if v"1.0" <= VERSION < v"1.1"
    _colon(I::CartesianIndex{N}, J::CartesianIndex{N}) where N =
        CartesianIndices(map((i,j) -> i:j, Tuple(I), Tuple(J)))
end
```

and replace `-ci1:ci1` with `_colon(-ci1:ci1)`

Reply:
In this case the `-ci1:ci1` construct is so much nicer that I'd agree we should introduce the compatibility code.

Reply:
You know, I'm going to add this to Compat.jl. Hold off a bit.

Reply:
I think it may take a while for Compat to percolate up (I had to remove lots of packages from my environment to allow me to `dev` it), so we should probably do our own thing for now.